csharp/src/Ice/OutputStream.cs

```diff
@@ -114,38 +108,35 @@ public Position(int segment, int offset)
         // segment and the offset into it.
         private Position _tail;
 
-        /// <summary>
-        /// This constructor uses the communicator's default encoding version.
-        /// </summary>
-        /// <param name="communicator">The communicator to use when initializing the stream.</param>
-        public OutputStream(Communicator communicator)
-            : this(communicator, communicator.DefaultsAndOverrides.DefaultEncoding)
-        {
-        }
+        internal Position Tail => _tail;
 
-        /// <summary>
-        /// This constructor uses the given communicator and encoding version.
-        /// </summary>
-        /// <param name="communicator">The communicator to use when initializing the stream.</param>
+        /// <summary>This constructor uses the given communicator and encoding version.</summary>
         /// <param name="encoding">The desired encoding version.</param>
-        /// <param name="buffer">The intial stream data.</param>
-        public OutputStream(Communicator communicator, Encoding encoding, byte[]? buffer = null)
+        /// <param name="data">The intial stream data.</param>
+        /// <param name="tail">The position at what start writing.</param>
+        internal OutputStream(Encoding encoding, List<ArraySegment<byte>> data, Position? tail = null)
```

Comment:
I think this constructor (or possibly another overloaded constructor) should tell the OutputSteam whether or not to start the main encapsulation. 

And the OutputSteam overall could/should be further simplified. For example, we don't need to set / reset mainEncaps anymore: for its entire lifetime an OutputStream writes to a main encaps or does not write to a main encaps.

Reply:
We will have to also pass the `FormatType` as a parameter here

Reply:
Yes.